### PR TITLE
Destinations CDK: CatalogParser sets default namespace

### PR DESCRIPTION
**Pull request by edgao**
* CatalogParser is now responsible for setting the default namespace on stream objects.
    * This entails a bunch of knock-on changes:
        * All the core async stuff now operates on the original (null) namespace
        * BufferEnqueue also needs to set the default namespace right before handing the message to the queue
        * AsyncStreamConsumer needs to set the default namespace when calling OnCloseFunction
    * I also took the opportunity to make defaultNamespace non-nullable.
* This allows destinations code to stop thinking about it (i.e. we can remove all our `catalog.streams.each(stream.setNamespace)` blocks)

closes https://github.com/airbytehq/airbyte-internal-issues/issues/2413

### CODE REVIEW

**Comment by edgao** [comment]
* **#38121** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/38121?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @edgao and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/integrations/destination/async/buffers/BufferEnqueue.kt [hunk 1]
```diff
@@ -53,15 +54,34 @@ class BufferEnqueue(
             }
         val stateId = stateManager.getStateIdAndIncrementCounter(streamDescriptor)
 
-        var addedToQueue = queue.offer(message, sizeInBytes.toLong(), stateId)
+        // We don't set the default namespace until after putting this message into the state
+        // manager/etc.
+        // All our internal handling is on the true (null) namespace,
+        // we just set the default namespace when handing off to destination-specific code.
+        val mangledMessage =
+            if (Strings.isNullOrEmpty(message.record!!.namespace)) {
```
Comment (gisripa) [inline airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/integrations/destination/async/buffers/BufferEnqueue.kt:62]: probably can use `namespace.isNullOrBlank()` extension function

File: airbyte-cdk/java/airbyte-cdk/db-destinations/src/main/kotlin/io/airbyte/cdk/integrations/destination/jdbc/AbstractJdbcDestination.kt [hunk 2]
```diff
@@ -245,22 +244,7 @@ abstract class AbstractJdbcDestination<DestinationState : MinimumDestinationStat
...
-                outputRecordCollector,
-                database,
-                sqlOperations,
-                namingResolver,
-                config,
-                catalog,
-                null,
-                NoopTyperDeduper(),
-            )
-        }
```
Comment (gisripa) [inline airbyte-cdk/java/airbyte-cdk/db-destinations/src/main/kotlin/io/airbyte/cdk/integrations/destination/jdbc/AbstractJdbcDestination.kt:260]: yay!. no more non-dv2 code.

Comment (edgao) [reply airbyte-cdk/java/airbyte-cdk/db-destinations/src/main/kotlin/io/airbyte/cdk/integrations/destination/jdbc/AbstractJdbcDestination.kt:260]: there's probably dozens of these scattered everywhere, but yeah we can start wiping them out

File: airbyte-cdk/java/airbyte-cdk/core/src/test/kotlin/io/airbyte/cdk/integrations/destination/async/AsyncStreamConsumerTest.kt [hunk 3]
```diff
@@ -146,9 +145,9 @@ class AsyncStreamConsumerTest {
                 onClose = onClose,
                 onFlush = flushFunction,
                 catalog = CATALOG,
-                bufferManager = BufferManager(),
+                bufferManager = BufferManager("default_ns"),
```
Comment (stephane-airbyte) [inline airbyte-cdk/java/airbyte-cdk/core/src/test/kotlin/io/airbyte/cdk/integrations/destination/async/AsyncStreamConsumerTest.kt:148]: I feel like there's still something funky here. We need to pass the default namespace twice. Seems to be an improvement though

Comment (edgao) [reply airbyte-cdk/java/airbyte-cdk/core/src/test/kotlin/io/airbyte/cdk/integrations/destination/async/AsyncStreamConsumerTest.kt:148]: yep. I really wish we had some sort of wrapper/replacement object for RecordMessage, so we could track `real_namespace: String?` alongside `nonnull_namespace: String`, and not need to worry about "this half of our codebase operates on nullable namespace and the other half operates on nonnullable"

b/c then we could just set the nonnull namespace immediately after reading the message off of stdin, and it's up to each component whether they want to use the real namespace or the nonnull version

Comment (stephane-airbyte) [reply airbyte-cdk/java/airbyte-cdk/core/src/test/kotlin/io/airbyte/cdk/integrations/destination/async/AsyncStreamConsumerTest.kt:148]: can we reference the namespace from the bufferManager object? That would be slightly cleaner

Comment (edgao) [reply airbyte-cdk/java/airbyte-cdk/core/src/test/kotlin/io/airbyte/cdk/integrations/destination/async/AsyncStreamConsumerTest.kt:148]: hm, not sure what you mean by that - buffer manager doesn't have a reference to anything that knows the default namespace, and I don't think we want to give it a reference to AsyncStreamConsumer?

(I _should_ switch this to use the DEFAULT_NAMESPACE constant though)

**Review by stephane-airbyte — APPROVED** [review]

**Review by gisripa — APPROVED** [review]

**Pull request merged**